Mathlib/Topology/Algebra/ProperAction.lean

```diff
@@ -0,0 +1,309 @@
+
+variable {G X Y Z : Type*} [Group G] [MulAction G X] [MulAction G Y]
+variable [TopologicalSpace G] [TopologicalSpace X] [TopologicalSpace Y]
+variable [TopologicalSpace Z]
+
+/-- If a group acts properly then in particularl it acts continuously. -/
+@[to_additive "If a group acts properly then in particularl it acts continuously."]
+instance continuousSmul_of_properSMul [ProperSMul G X] : ContinuousSMul G X where
+  continuous_smul := (isProperMap_smul_pair G X).continuous.fst
+
+/-- A group `G` acts properly on a topological space `X` if and only if for all ultrafilters
+`𝒰` on `X × G`, if `𝒰` converges to `(x₁, x₂)` along the map `(g, x) ↦ (g • x, x)`,
+then there exists `g : G` such that `g • x₂ = x₁` and `𝒰.fst` converges to `g`. -/
+@[to_additive "A group acts `G` properly on a topological space `X` if and only if
+for all ultrafilters `𝒰` on `X`, if `𝒰` converges to `(x₁, x₂)`
+along the map `(g, x) ↦ (g • x, x)`, then there exists `g : G` such that `g • x₂ = x₁`
+and `𝒰.fst` converges to `g`."]
+theorem properSMul_iff_continuousSMul_ultrafilter_tendsto : ProperSMul G X ↔ ContinuousSMul G X ∧
+    (∀ 𝒰 : Ultrafilter (G × X), ∀ x₁ x₂ : X,
+    Tendsto (fun gx ↦ ⟨gx.1 • gx.2, gx.2⟩ : G × X → X × X) 𝒰 (𝓝 (x₁, x₂)) →
+    ∃ g : G, g • x₂ = x₁ ∧ Tendsto Prod.fst (𝒰 : Filter (G × X)) (𝓝 g)) := by
+  constructor
+  · refine fun h ↦ ⟨inferInstance, fun 𝒰 x₁ x₂ h' ↦ ?_⟩
+    rw [properSMul_iff, isProperMap_iff_ultrafilter] at h
+    rcases h.2 h' with ⟨gx, hgx1, hgx2⟩
+    refine ⟨gx.1, ?_, (continuous_fst.tendsto gx).mono_left hgx2⟩
+    simp only [Prod.mk.injEq] at hgx1
+    rw [← hgx1.2, hgx1.1]
+  · rintro ⟨cont, h⟩
+    rw [properSMul_iff, isProperMap_iff_ultrafilter]
+    refine ⟨by fun_prop, fun 𝒰 (x₁, x₂) hxx ↦ ?_⟩
+    rcases h 𝒰 x₁ x₂ hxx with ⟨g, hg1, hg2⟩
+    refine ⟨(g, x₂), by rw [hg1], ?_⟩
+    rw [nhds_prod_eq, 𝒰.le_prod]
+    exact ⟨hg2, (continuous_snd.tendsto _).comp hxx⟩
+
+/-- A group `G` acts properly on a T2 topological space `X` if and only if for all ultrafilters
+`𝒰` on `X × G`, if `𝒰` converges to `(x₁, x₂)` along the map `(g, x) ↦ (g • x, x)`,
+then there exists `g : G` such that `𝒰.fst` converges to `g`. -/
+theorem properSMul_iff_continuousSMul_ultrafilter_tendsto_t2 [T2Space X] : ProperSMul G X ↔
+    ContinuousSMul G X ∧
+    (∀ 𝒰 : Ultrafilter (G × X), ∀ x₁ x₂ : X,
+    Tendsto (fun gx ↦ ⟨gx.1 • gx.2, gx.2⟩ : G × X → X × X) 𝒰 (𝓝 (x₁, x₂)) →
+    ∃ g : G, Tendsto Prod.fst (𝒰 : Filter (G × X)) (𝓝 g)) := by
+  rw [properSMul_iff_continuousSMul_ultrafilter_tendsto]
+  refine and_congr_right fun hc ↦ ?_
+  congrm ∀ 𝒰 x₁ x₂ hxx, ∃ g, ?_
+  exact and_iff_right_of_imp fun hg ↦ tendsto_nhds_unique
+    (hg.smul ((continuous_snd.tendsto _).comp hxx)) ((continuous_fst.tendsto _).comp hxx)
+
+/-- If `G` acts properly on `X`, then the quotient space is Hausdorff (T2). -/
+@[to_additive "If `G` acts properly on `X`, then the quotient space is Hausdorff (T2)."]
+theorem t2Space_of_ProperSMul (hproper:ProperSMul G X) :
```

Comment:
Since you're making `ProperSMul` a typeclass, you should use it as a typeclass argument
```suggestion
theorem t2Space_of_ProperSMul [ProperSMul G X] :
```